pkg/event/registry_test.go

```diff
@@ -17,95 +17,91 @@ const npOVNKubernetes = "OVNKubernetes"
-			handler2 = NewTestHandler("handler2", npGenericKubeproxyIptables)
-			err := registry.AddHandlers(logger.NewHandler(), handler1, handler2)
+			err := registry.AddHandlers(logger.NewHandler(), matchingHandlers[0], nonMatchingHandlers[0], matchingHandlers[1],
+				matchingHandlers[2])
 			Expect(err).NotTo(HaveOccurred())
-			cleanupTestEvents()
 		})
 
-		It("should initialize all the handlers", func() {
-			Expect(handler1.Initialized).To(BeTrue())
-			Expect(handler2.Initialized).To(BeTrue())
-		})
+		It("should initialize all matching handlers", func() {
+			for _, h := range matchingHandlers {
+				Expect(h.Initialized).To(BeTrue())
+			}
 
-		It("should invoke all handlers for all events", func() {
-			endpoint := &submV1.Endpoint{ObjectMeta: v1meta.ObjectMeta{Name: "endpoint1"}}
-			node := &k8sV1.Node{ObjectMeta: v1meta.ObjectMeta{Name: "node1"}}
-
-			events := map[TestEvent]func() error{
-				{Name: evStop, Parameter: false}:                     func() error { return registry.StopHandlers(false) },
-				{Name: evTransitionToGateway}:                        registry.TransitionToGateway,
-				{Name: evTransitionToNonGateway}:                     registry.TransitionToNonGateway,
-				{Name: evNodeCreated, Parameter: node}:               func() error { return registry.NodeCreated(node) },
-				{Name: evNodeUpdated, Parameter: node}:               func() error { return registry.NodeUpdated(node) },
-				{Name: evNodeRemoved, Parameter: node}:               func() error { return registry.NodeRemoved(node) },
-				{Name: evLocalEndpointCreated, Parameter: endpoint}:  func() error { return registry.LocalEndpointCreated(endpoint) },
-				{Name: evLocalEndpointUpdated, Parameter: endpoint}:  func() error { return registry.LocalEndpointUpdated(endpoint) },
-				{Name: evLocalEndpointRemoved, Parameter: endpoint}:  func() error { return registry.LocalEndpointRemoved(endpoint) },
-				{Name: evRemoteEndpointCreated, Parameter: endpoint}: func() error { return registry.RemoteEndpointCreated(endpoint) },
-				{Name: evRemoteEndpointUpdated, Parameter: endpoint}: func() error { return registry.RemoteEndpointUpdated(endpoint) },
-				{Name: evRemoteEndpointRemoved, Parameter: endpoint}: func() error { return registry.RemoteEndpointRemoved(endpoint) },
+			for _, h := range nonMatchingHandlers {
+				Expect(h.Initialized).To(BeFalse())
 			}
+		})
 
-			for ev, f := range events {
+		It("should invoke all matching handlers of all events", func() {
+			for ev, f := range allEvents(registry) {
 				err := f()
 				Expect(err).To(Succeed())
-				Expect(handler1.Events).Should(Receive(Equal(ev)))
-				Expect(handler2.Events).Should(Receive(Equal(ev)))
-
-				// Handlers should be handled invoked in order of registration
-				ev.Handler = handler1.Name
-				Expect(allTestEvents).Should(Receive(Equal(ev)))
-				ev.Handler = handler2.Name
-				Expect(allTestEvents).Should(Receive(Equal(ev)))
+
+				for _, h := range matchingHandlers {
```

Comment:
In this unit testing it's not implicit that we expect that the handlers should be called in order. I'd add a comment at least.

Imagining that someone rewrites the registry (somehow.. with a map...) .. and then the order is broken, so they change this test here.

If we say explicitly somewhere here that the events of the handlers should be received in order ... it'd safer I think

Reply:
yeah - that is already described in the registry `AddHandlers` API - "_Handlers will be called in registration order"_. But I add it to the _It_ text as well.